docs/cancellation.md

```diff
@@ -0,0 +1,439 @@
+# Cancellation in Hyperlight
+
+This document describes the cancellation mechanism and memory ordering guarantees for Hyperlight.
+
+## Overview (Linux)
+
+Hyperlight provides a mechanism to forcefully interrupt guest execution through the `InterruptHandle::kill()` method. This involves coordination between multiple threads using atomic operations and POSIX signals to ensure safe and reliable cancellation.
+
+## Key Components
+
+### LinuxInterruptHandle State
+
+The `LinuxInterruptHandle` uses a packed atomic u64 to track execution state:
+
+- **state (AtomicU64)**: Packs two bits:
+  - **Bit 1 (RUNNING_BIT)**: Set when vCPU is actively running in guest mode
+  - **Bit 0 (CANCEL_BIT)**: Set when cancellation has been requested via `kill()`
+- **tid (AtomicU64)**: Thread ID where the vCPU is running
+- **debug_interrupt (AtomicBool)**: Set when debugger interrupt is requested (gdb feature only)
+- **dropped (AtomicBool)**: Set when the corresponding VM has been dropped
+
+The packed state enables atomic reads of both RUNNING_BIT and CANCEL_BIT simultaneously via `get_running_and_cancel()`. Within a single `run()` call, the CANCEL_BIT remains set across vcpu exits and re-entries (such as when calling host functions), ensuring cancellation persists until the guest call completes. However, `clear_cancel()` resets the CANCEL_BIT at the beginning of each new `run()` call, preventing cancellation requests from affecting subsequent guest function calls.
+
+### Signal Mechanism
+
+On Linux, Hyperlight uses `SIGRTMIN + offset` (configurable, default offset is typically 0) to interrupt the vCPU thread. The signal handler is intentionally a no-op - the signal's only purpose is to cause a VM exit via `EINTR` from the `ioctl` call that runs the vCPU.
+
+## Run Loop Flow
+
+The main execution loop in `HyperlightVm::run()` coordinates vCPU execution with potential interrupts. Here's the detailed flow:
+
+```mermaid
+sequenceDiagram
+    participant VM as run() Loop
+    participant Guest as vCPU (Guest)
+    participant IH as InterruptHandle
+
+    Note over VM: === TIMING POINT 1 ===
+    VM->>IH: clear_cancel()
+    Note right of VM: Clear premature kill()<br/>requests
+
+    loop Run Loop
+        Note over VM: === TIMING POINT 2 ===
+        VM->>IH: set_tid()
+        Note right of VM: Store current thread ID
+        VM->>IH: set_running()
+        Note right of VM: Set running=true
+
+        VM->>IH: is_cancelled()
+        
+        alt is_cancelled() == true
```

Comment:
These `alt` seem like they could be block names in the diagrams which would make it more readable: 
<img width="1362" height="252" alt="image" src="https://github.com/user-attachments/assets/8db67a22-2353-400a-8644-b5da7b2acab4" />

Reply:
The block name are located next to the alts. In this case `[is_cancelled()==true] `